### ISSUE TITLE
twisted.web.xmlrpc._QueryFactory.parseResponse fails to errback for a particular malformed response

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3399]
|<img alt="toranin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| toranin reported|
|-|-|
|Trac ID|trac#3399|
|Type|defect|
|Created|2008-08-27 19:04:43Z|
|Branch|https://github.com/twisted/twisted/tree/xmlrpc-empty-response-3399|

(applies to 8.1.0, maybe newer)

If the server returns XML that parses but contains no data, an `IndexError` is thrown from line 321, in the else part of the try/except block that is supposed to catch problems with the response structure.

I moved the indexing operation in question into the try block in the attached patch.

Attachments:

* [twisted.web.xmlrpc-8.1.0-fix.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d0d/d0d9d29f2e4b1f5cbcd2018298f74bd3121ba49a/106b4bcc5ed7671643224e191f8c8a43060428e8.patch) (650 bytes) - added by toranin on 2008-08-27 19:06:21Z - Fix for this issue in twisted.web.xmlrpc in version 8.1.0

<details><summary>Searchable metadata</summary>

```
trac-id__3399 3399
type__defect defect
reporter__toranin toranin
priority__normal normal
milestone__ 
branch__branches_xmlrpc_empty_response_3399 branches/xmlrpc-empty-response-3399
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1219863883000000 1219863883000000
changetime__1220606314000000 1220606314000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

What response provokes this?

**Comment by twisted-trac** [comment]
|<img alt="toranin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|toranin commented|
|-|-|

Here's the response:
```
Date: Wed, 27 Aug 2008 19:24:45 UTC
Content-Type: text/xml
Connection: close
Cache-control: private
Content-Length: 99

<?xml version="1.0"?>
<methodResponse>
 <params>
  <param>
  </param>
 </params>
</methodResponse>
```
It's produced by an in-house server with a couple of bugs, and that will be fixed at some point.  Nonetheless, twisted's xmlrpc handler ought not to 'hang' when presented with this.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [24664]) Branching to 'xmlrpc-empty-response-3399'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [24666]) Apply the fix, with a test.

Refs [#3399](https://github.com/twisted/twisted/issues/3399)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

Please review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

The comment in `test_parseResponseWithoutData` isn't quite right.  The failure mode is that it `parseResponse` raises `IndexError`, right?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

Hum I got confused. It should be better now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

Looks better.  `twisted.web.xmlrpc.Proxy.callRemote` could perhaps use some documentation regarding its errors (well, it has no documentation at all right now).  This change adds a new way it can fail, so it would be good to at least mention that.

Merge when you're happy with it.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set status to `closed`|
|-|-|

(In [24679]) Merge xmlrpc-empty-response-3399

Author: therve
Reviewer: exarkun
Fixes [#3399](https://github.com/twisted/twisted/issues/3399)

Fix xmlrpc client so that he manages to return an error for buggy servers
sending an empty response.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
